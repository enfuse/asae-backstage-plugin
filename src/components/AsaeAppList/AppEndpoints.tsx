
import { Box, Chip, Collapse, Paper } from '@mui/material'
import { AsaeApp } from '../../types'
import React from 'react'
import { useStyles } from "../common/styles";
import { DataProps } from '../common/withAzureClient';

export const  AppEndpoints = ( {asaeApp}:any) =>{
    return (
      <Paper elevation={6} sx={{background: asaeApp.endpoints.length > 0? 'default':'e6e8e6'}} >
        <Box sx={{ padding: 1, paddingLeft:4 }}>
          <h2>Endpoints</h2>
          {asaeApp?.endpoints?.map(endpoint=>
            <Endpoint endpoint={endpoint} ></Endpoint>
          )}
        </Box>
    </Paper>
    )
  }

const Endpoint = (props : {endpoint}) => {
    const {endpoint} = props 
    const headers = endpoint?.details?.requestMappingConditions?.headers
    const params = endpoint?.details?.requestMappingConditions?.params
    const consumes = endpoint?.details?.requestMappingConditions?.consumes
    const produces = endpoint?.details?.requestMappingConditions?.produces
    const [open, setOpen] = React.useState(false)
    const classes = useStyles()
    console.log(produces)
    return (
        <Box >
        <p onClick={()=>setOpen(!open)} className={classes.rowClass}>
          <b> <EndpointMethod  method={endpoint?.details?.requestMappingConditions?.methods[0]} /> </b>
        {endpoint?.details?.requestMappingConditions?.patterns[0]}
        </p>
        {/* <Collapse in={open} timeout="auto" unmountOnExit sx={{padding:0, margin:0}}>
            <Box sx={{ paddingLeft:4}}>

            <p><b>Headers&nbsp;</b>
            {headers.map(header => <>{header}&nbsp;</>)}
            </p>
            <p> <b>Params&nbsp;</b></p>
            {params.map(param => {
                <p>{param}</p>
            })}
            <p><b>Consumes&nbsp;</b></p>
            {consumes.map(consumeItem => {
                <p>{consumeItem}</p>
            })}
            <p><b>Produces&nbsp;</b>
            {produces.map(produceItem =>  <>{produceItem.mediaType}&nbsp;</>)}
            </p>
            </Box>
        </Collapse> */}
      </Box>
    )

}

const EndpointMethod = (props :{method:string}) => {
    const {method} = props
    return (
        <>
        {
            {
                'GET': <Chip label="GET" color="success"></Chip>,
                'POST': <Chip label="POST" color="primary"></Chip>,
                'PUT': <Chip label="PUT" color="info"></Chip>,
                'DELETE': <Chip label="DELETE" color="error" ></Chip>
            }[method]
        }
        </>
    )
}