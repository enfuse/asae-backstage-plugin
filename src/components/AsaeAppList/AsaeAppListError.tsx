import React from 'react'
import Alert from '@material-ui/lab/Alert';
import { AlertTitle } from '@material-ui/lab';

type Error = {
  name: string;
  message: string;
  stack?: string;
};

export const AsaeAppListError = (props : {error :Error} ) => {
    return <Alert severity="error" variant='outlined'>
            <AlertTitle>Error in ASAE App list component</AlertTitle>
            {props.error.message}
          </Alert>
}