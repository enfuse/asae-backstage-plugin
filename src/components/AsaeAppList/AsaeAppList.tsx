import React from 'react'
// import { TableColumn, Table } from '@backstage/core-components';
import useAsync from 'react-use/lib/useAsync';
import { AppPlatformManagementClient,AppResource} from "@azure/arm-appplatform";

import { withAzureClient,DataProps} from '../common/withAzureClient';
import * as Types  from '../../types'
import {AsaeApp} from '../../types'
import * as Utils from '../../utils/asae-utils'
import { AsaeAppListError } from './AsaeAppListError';
import {AppEndpoints} from './AppEndpoints'

import Box from '@mui/material/Box';
import Collapse from '@mui/material/Collapse';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import {StatusError,StatusOK } from '@backstage/core-components';
import Paper from '@mui/material/Paper';
import { useStyles } from "../common/styles";
import { Grid } from '@material-ui/core';
import axios from 'axios';





const isActuatorMapping = (mapping)=>{
    const handlerClassName : string = mapping?.details?.handlerMethod?.className
    if (!!handlerClassName){
        return handlerClassName.includes('org.springframework')
    }
    return true
}
const fetchEndpoints  = async (url:string) => {

    const endpoints = []
    try{
        const response = await axios.get(url.concat('/actuator/mappings'))
        const mappings = response?.
                        data?.contexts?.application?.mappings
                        ?.dispatcherServlets?.dispatcherServlet
        for(let mapping  of mappings){
            if(!isActuatorMapping(mapping)) endpoints.push(mapping)
        }}
    catch(e:any){console.log(e)}
    return endpoints
}

function Row( props: {asaeApp: AsaeApp} ) {
  const {asaeApp} = props
  const [open, setOpen] = React.useState(false);
  const classes = useStyles()  
  return (
    <React.Fragment>
      <TableRow className={classes.rowClass}  onClick={()=>setOpen(!open) }> 
        <TableCell sx={{border:0}}>{asaeApp.appName}</TableCell >
        <TableCell sx={{border:0}}>{asaeApp?.state}</TableCell >
        <TableCell sx={{border:0}} >{asaeApp.provisioningState}</TableCell>
        <TableCell sx={{border:0}}>{asaeApp.location}</TableCell>
      </TableRow>
      <TableRow className={classes.collapsedContent}>
        <TableCell sx={{border:0}} style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Paper>
              <Box sx={{ padding: 1 }}>
                <Grid container spacing={2}>
                  <Grid item xs={12} >
                    <Paper elevation={6}>
                      <Box sx={{ padding: 2, paddingLeft:4}}>
                        <h2><i>{asaeApp.appName}</i></h2>
                        <Grid container spacing={5}>
                            <Grid item xs={6}>
                              <div>
                                <p><b>Domain </b>{asaeApp?.properties?.fqdn}</p>
                                <p><b>URL </b>{asaeApp?.properties?.url}</p>
                              </div>
                            </Grid>
                            <Grid item xs={6} >
                              <div>
                                <p>{asaeApp?.properties?.enableEndToEndTLS? <StatusOK>END-TO-END-TLS</StatusOK>:<StatusError>END-TO-END-TLS</StatusError>}</p>
                                <p>{asaeApp?.properties?.httpsOnly? <StatusOK>HTTPS-ONLY</StatusOK>:<StatusError>HTTPS-ONLY</StatusError>}</p>
                                <p>{asaeApp?.properties?.public? <StatusOK>PUBLIC</StatusOK>:<StatusError>PUBLIC</StatusError>}</p>
                              </div>
                            </Grid>
                        </Grid>
                      </Box>
                    </Paper>
                  </Grid>
                  <Grid item xs={12} sm={12} md={6}>
                      <Paper elevation={6}>
                          <Box sx={{ padding: 1, paddingLeft:4 }}>
                              <h2>Storage</h2>
                      <Grid container spacing={5}>
                          <Grid item xs={6}>
                          <div>
                              <h3>Temporary</h3>
                              <p><b>Mount Path</b> {asaeApp?.properties?.temporaryDisk?.mountPath}</p>
                              <p><b>Size </b>{asaeApp?.properties?.temporaryDisk?.sizeInGB}Gb</p>
                          </div>
                          </Grid>
                          <Grid item xs={6} >
                          <div>
                              <h3>Persistent</h3>
                              <p><b>Mount Path</b> {asaeApp?.properties?.persistentDisk?.mountPath}</p>
                              <p><b>Size </b>{asaeApp?.properties?.persistentDisk?.sizeInGB}Gb</p>
                          </div>
                          </Grid>
                      </Grid>
                          </Box>
                      </Paper>
                  </Grid>
                  <Grid item xs={12} sm={12} md={6}>
                      <Paper elevation={6} sx={{background : !asaeApp?.identity?.principalId ? '#e6e8e6' : 'default'}}>
                          <Box sx={{ padding: 1, paddingLeft:4 }}>
                            {!asaeApp?.identity?.principalId && <p className={classes.notPresent}>Identify Information not provided</p>}
                            <h2>Identity</h2>
                            <div role="identity-data">
                              <p><b>Principal ID </b>{asaeApp?.identity?.principalId}</p>     
                              <p><b>Tenant ID </b>{asaeApp?.identity?.tenantId}</p>
                              <p><b>Type </b>{asaeApp?.identity?.type}</p>
                            </div>
                          </Box>
                      </Paper>
                  </Grid>
                  <Grid item xs={12} sm={12} md={6}>
                        <AppEndpoints asaeApp={asaeApp}> </AppEndpoints>
                  </Grid>
                </Grid>
              </Box> 
            </Paper>
          </Collapse>
        </TableCell>
      </TableRow> 
      </React.Fragment>
  );
}

export const AsaeAppsList =({value}:DataProps) =>{
  return (
    <TableContainer component={Paper}>
      <Table aria-label="collapsible table" >
        <TableHead>
          <TableRow>
            <TableCell>App Name</TableCell>
            <TableCell> App State</TableCell>
            <TableCell >Provisioning State</TableCell>
            <TableCell >Location</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {value.map((item:AsaeApp) => (
            <Row key={item.id} asaeApp={item} />
          ))}
        </TableBody>
      </Table>
    </TableContainer>

  );
}


export const asaeAppListCallback = (client : AppPlatformManagementClient, entityAsaeInfo: Types.EntityAsae)=>{
    const { value, loading, error } = useAsync(async (): Promise<AppResource[]> => {
        const asaeApps: AsaeApp[]  = []

        const appsInterator = client.apps.list(entityAsaeInfo.resourceGroup, entityAsaeInfo.asaeService)
        for await (const app of appsInterator) { 
          console.log("appresource: ",app)
            asaeApps.push({
                id: app.id,
                appName:app.name,
                location:app.location,
                provisioningState: Utils.getStatusComponent(app?.properties?.provisioningState),
                identity: app.identity,
                properties: app.properties,
                state: await Utils.getApplicationState(app?.properties?.url),
                endpoints: await fetchEndpoints(app?.properties?.url)
            } as AsaeApp) 
        }

        return await asaeApps
      },[]);
      return { value, loading, error }
}

export const AsaeBuildPacksWithAzureClient = withAzureClient(AsaeAppsList, 
    AsaeAppListError,
    asaeAppListCallback)