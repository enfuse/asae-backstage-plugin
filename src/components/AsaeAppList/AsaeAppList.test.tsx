import React from 'react';
import { render, screen } from '@testing-library/react';
import { AsaeAppsList } from './AsaeAppList';
import userEvent from '@testing-library/user-event'
describe('When DataTable renders', () => {
  let rendered : HTMLElement | undefined = undefined 

  const setup = () => {

      const testData = [{
        appName: 'candidate-service-dev',
        provisioningState: 'Successful',
        location: 'us-west2',
        identity: {tenantId: 'tenantId-test-1', clientId: 'clientId-test-1'},
        id: 1
      },{
        appName: 'candidate-portal-dev',
        provisioningState: 'Pending',
        location: 'us-east1',
        identity: {tenantId: '', clientId: ''},
        id: 2
        },
      ]

      const {baseElement} =  render(<AsaeAppsList value={testData}/>)
      rendered = baseElement
  }
  beforeEach(()=>{
    setup()
  })

  it('should display 3 column headers', async() => {
      expect(screen.getAllByRole('columnheader').length).toBe(3)
      // expect(screen.getAllByRole('columnheader')[0]).toBe('test')
  })
  it('display table with data', async()=>{    
      expect(screen.getByRole('cell', {name:"candidate-service-dev"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"candidate-portal-dev"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"Successful"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"Pending"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"us-west2"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"us-east1"})).toBeInTheDocument()
  })
  it ('should display Identity information', ()=>{
    // TODO figure out how to test
    // fireEvent.click(screen.getAllByRole('row')[1])
    userEvent.click(screen.getAllByRole('row')[1])
    console.log(screen.getAllByRole('row')[1].outerHTML)
    console.log(screen.debug())

    expect(screen.getByRole('identity-data')).toBeInTheDocument()
  })

})
