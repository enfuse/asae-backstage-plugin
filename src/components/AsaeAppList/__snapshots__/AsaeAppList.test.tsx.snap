// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When DataTable renders should match previous snapshot 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="collapsible table"
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              App Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Provisioning State
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Location
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root makeStyles-rowClass-7 css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.appname-1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.provisioningState-1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.location-1
            </td>
          </tr>
          <tr
            class="MuiTableRow-root makeStyles-collapsedContent-8 css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
              colspan="6"
              style="padding-bottom: 0px; padding-top: 0px;"
            />
          </tr>
          <tr
            class="MuiTableRow-root makeStyles-rowClass-7 css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.appname-2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.provisioningState-2
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
            >
              item.location-2
            </td>
          </tr>
          <tr
            class="MuiTableRow-root makeStyles-collapsedContent-8 css-1q1u3t4-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1q8rm1h-MuiTableCell-root"
              colspan="6"
              style="padding-bottom: 0px; padding-top: 0px;"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
